File: Hot100/src/com/action/hash/GroupAnagrams.java
package com.action.hash;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class GroupAnagrams {

    /*
    source: https://leetcode.com/problems/group-anagrams
    desc:
        给你一个字符串数组，请你将 字母异位词 组合在一起。可以按任意顺序返回结果列表。
    * */

    public static List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> map = new HashMap<>();
        // 字母异位词使用同一个key：排序、字符拼接
        for(String s:strs){
            String key = getKey(s);
            List<String> value = map.getOrDefault(key, new ArrayList<>());
            value.add(s);
            map.put(key, value);
        }
        return new ArrayList<>(map.values());
    }

    private static String getKey(String s){
        int[] cnt = new int[26];
        int len = s.length();
        for(int i=0; i<len; i++){
            cnt[s.charAt(i)-'a']++;
        }
        StringBuilder sb = new StringBuilder();
        for(int i=0; i<26; i++){
            if(cnt[i] == 0) continue;
            sb.append(cnt[i]);
            sb.append((char)('a'+i));
        }
        return sb.toString();
    }


    public static void main(String[] args) {
        String[] strs = new String[]{"eat","tea","tan","ate","nat","bat"};
        List<List<String>> result = groupAnagrams(strs);
        System.out.println(result);
    }
}

File: Go/Hot100/hash/groupAnagrams.go
package hash

// 给你一个字符串数组，请你将 字母异位词 组合在一起。可以按任意顺序返回结果列表。
//
// 示例 1:
//
// 输入: strs = ["eat", "tea", "tan", "ate", "nat", "bat"]
//
// 输出: [["bat"],["nat","tan"],["ate","eat","tea"]]
//
// 解释：
//
// 在 strs 中没有字符串可以通过重新排列来形成 "bat"。
// 字符串 "nat" 和 "tan" 是字母异位词，因为它们可以重新排列以形成彼此。
// 字符串 "ate" ，"eat" 和 "tea" 是字母异位词，因为它们可以重新排列以形成彼此。
func GroupAnagrams(strs []string) [][]string {
	hashMap := make(map[[26]int][]string)
	for _, str := range strs {
		key := [26]int{}
		for _, c := range str {
			key[c-'a']++
		}
		hashMap[key] = append(hashMap[key], str)
	}
	result := make([][]string, 0, len(hashMap))
	for _, v := range hashMap {
		result = append(result, v)
	}
	return result
}
